File: Lab9/app/src/main/java/com/example/lab9/MainActivity.java
package com.example.lab9;

import androidx.appcompat.app.AppCompatActivity;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import java.util.Locale;

public class MainActivity extends AppCompatActivity
{
    private TextView tvClock;
    private Button btnStart;
    private Boolean flag = false;
    private BroadcastReceiver receiver = new BroadcastReceiver()
    {
        @Override
        public void onReceive(Context context, Intent intent)
        {
            Bundle b = intent.getExtras();
            tvClock.setText(String.format(Locale.getDefault(), "%02d:%02d:%02d", b.getInt("H"), b.getInt("M"), b.getInt("S")));
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        tvClock = findViewById(R.id.tvClock);
        btnStart = findViewById(R.id.btnStart);

        registerReceiver(receiver, new IntentFilter("MyMessage"));

        flag = MyService.flag;

        if(flag)
        {
            btnStart.setText("Pause");
        }
        else
        {
            btnStart.setText("Start");
        }

        btnStart.setOnClickListener(v ->
        {
            flag = !flag;

            if(flag)
            {
                btnStart.setText("Pause");
                Toast.makeText(this, "Timing Start", Toast.LENGTH_SHORT).show();
            }
            else
            {
                btnStart.setText("Start");
                Toast.makeText(this, "Timing Pause", Toast.LENGTH_SHORT).show();
            }

            startService(new Intent(this, MyService.class).putExtra("flag", flag));
        });
    }
    @Override
    protected void onDestroy()
    {
        super.onDestroy();
        unregisterReceiver(receiver);
    }
}

File: Assignment6/KotlinContentProviderSub/app/src/main/java/com/example/kotlincontentprovidersub/MainActivity.kt
package com.example.kotlincontentprovidersub

import android.content.ContentValues
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.*

class MainActivity : AppCompatActivity() {
    private var items: ArrayList<String> = ArrayList()
    private lateinit var adapter: ArrayAdapter<String>
    private val uri = Uri.parse("content://com.example.kotlincontentprovidersub")
    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, items)
        findViewById<ListView>(R.id.listView).adapter = adapter
        setListener()
    }

    private fun setListener()
    {
        val edBook = findViewById<EditText>(R.id.edBook)
        val edPrice = findViewById<EditText>(R.id.edPrice)
        findViewById<Button>(R.id.btnInsert).setOnClickListener{
            val name = edBook.text.toString()
            val price = edPrice.text.toString()
            if (name.isEmpty() || price.isEmpty())
                showToast("Title/Price can't be empty!")
            else {
                val values = ContentValues()
                values.put("book", name)
                values.put("price", price)
                val contentUri = contentResolver.insert(uri, values)
                if (contentUri != null) {
                    showToast("Add $name, Price:$price")
                    cleanEditText()
                } else
                    showToast("Failed to add")
            }
        }
        findViewById<Button>(R.id.btnUpdate).setOnClickListener{
            val name = edBook.text.toString()
            val price = edPrice.text.toString()
            if (name.isEmpty() || price.isEmpty())
                showToast("Title/Price can't be empty")
            else {
                val values = ContentValues()
                values.put("price", price)
                val count = contentResolver.update(uri, values, name, null)
                if (count > 0)
                {
                    showToast("Update$name, Price:$price")
                    cleanEditText()
                }
                else
                    showToast("Failed to update")
            }
        }
        findViewById<Button>(R.id.btnDelete).setOnClickListener {
            val name = edBook.text.toString()
            if (name.isEmpty())
                showToast("Title can't be empty")
            else
            {
                val count = contentResolver.delete(uri, name, null)
                if (count > 0)
                {
                    showToast("Delete:${name}")
                    cleanEditText()
                }
                else
                    showToast("Failed to delete")
            }
        }
        findViewById<Button>(R.id.btnQuery).setOnClickListener {
            val name = edBook.text.toString()
            val selection = if (name.isEmpty()) null else name
            val c = contentResolver.query(uri, null, selection, null, null)
            c ?: return@setOnClickListener
            c.moveToFirst()
            items.clear()
            showToast("${c.count} results")
            for (i in 0 until c.count) {
                items.add("Title:${c.getString(0)}\t\t\t\t Price:${c.getInt(1)}")
                c.moveToNext()
            }
            adapter.notifyDataSetChanged()
            c.close()
        }
    }
    private fun showToast(text: String) =
        Toast.makeText(this,text, Toast.LENGTH_LONG).show()
    private fun cleanEditText() {
        findViewById<EditText>(R.id.edBook).setText("")
        findViewById<EditText>(R.id.edPrice).setText("")
    }
}